File: backend/rag/query.js
import { vectorStore } from "../vector/store.js";
import { embed } from "../llm/embedding.js";
import { cosineSimilarity } from "../utils/similarity.js";

export async function retrieveContext(question, k = 3) {
  const qEmbedding = await embed(question);

  const scored = vectorStore.map((item) => ({
    text: item.text,
    score: cosineSimilarity(qEmbedding, item.embedding),
  }));

  return scored
    .sort((a, b) => b.score - a.score)
    .slice(0, k)
    .map((x) => x.text)
    .join("\n");
}

File: backend/rag/query.py
from llm.embedding import embed
from vector.chroma import get_collection 
from utils.similarity import cosine_similarity  

async def retrieve_context(question, k=3):
    try:
        q_embedding = await embed(question)
        collection = await get_collection()

       
        results = collection.query(
            query_embeddings=[q_embedding],
            n_results=k,
            include=["documents", "metadatas", "distances"]  
        )

       
        top_texts = results["documents"][0] if results["documents"] else []
        if not top_texts:
            return ""

      
        context = "\n".join(top_texts)
        print(f"Top scores (distances): {results['distances'][0]}")  # Log để debug
        return context
    except Exception as error:
        print(f"❌ Lỗi retrieve: {error}")
        return ""